### ISSUE TITLE
Large amount of no such file or directory exceptions

### ISSUE THREAD

**Issue by bored** [RipMeApp/ripme#610]
* Ripme version: 1.7.47
* Java version:
```
openjdk version "1.8.0_171"
OpenJDK Runtime Environment (build 1.8.0_171-8u171-b11-1~deb9u1-b11)
OpenJDK 64-Bit Server VM (build 25.171-b11, mixed mode)
```
* Operating system: Debian GNU/Linux 9.4 (stretch)
* Exact URL you were trying to rip when the problem occurred: http://momentummod.tumblr.com
* Please include any additional information about how to reproduce the problem: Whenever I try to rip from Tumblr or Instagram (possibly others too) my console is filled with `java.io.IOException`s.

## Expected Behavior

For no exceptions to occur.

## Actual Behavior

```java
[...]
java.io.IOException: No such file or directory
	at java.io.UnixFileSystem.createFileExclusively(Native Method)
	at java.io.File.createNewFile(File.java:1012)
	at com.rarchives.ripme.ripper.AbstractRipper.writeDownloadedURL(AbstractRipper.java:77)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:256)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:265)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:269)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:283)
	at com.rarchives.ripme.ripper.AlbumRipper.addURLToDownload(AlbumRipper.java:112)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.handleJSON(TumblrRipper.java:237)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.rip(TumblrRipper.java:191)
	at com.rarchives.ripme.App.rip(App.java:92)
	at com.rarchives.ripme.App.ripURL(App.java:250)
	at com.rarchives.ripme.App.handleArguments(App.java:237)
	at com.rarchives.ripme.App.main(App.java:67)
[...]
```

<details>
	<summary>Full Output</summary>

```
~/Pictures
❯ java -jar ripme.jar -u http://momentummod.tumblr.com
Loaded file:/home/user/Pictures/ripme.jar!/rip.properties
Loaded log4j.properties
Initialized ripme v1.7.47
[+] Creating directory: ./rips/tumblr_momentummod.tumblr.com
Using api key: qpdkY6nMknksfvYAhf2xIHp0iNRLkMlcWShxqzXyFJRxIsZ1Zz
Retrieving http://api.tumblr.com/v2/blog/momentummod.tumblr.com/posts/photo?api_key=qpdkY6nMknksfvYAhf2xIHp0iNRLkMlcWShxqzXyFJRxIsZ1Zz&offset=0
java.io.IOException: No such file or directory
	at java.io.UnixFileSystem.createFileExclusively(Native Method)
	at java.io.File.createNewFile(File.java:1012)
	at com.rarchives.ripme.ripper.AbstractRipper.writeDownloadedURL(AbstractRipper.java:77)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:256)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:265)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:269)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:283)
	at com.rarchives.ripme.ripper.AlbumRipper.addURLToDownload(AlbumRipper.java:112)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.handleJSON(TumblrRipper.java:237)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.rip(TumblrRipper.java:191)
	at com.rarchives.ripme.App.rip(App.java:92)
	at com.rarchives.ripme.App.ripURL(App.java:250)
	at com.rarchives.ripme.App.handleArguments(App.java:237)
	at com.rarchives.ripme.App.main(App.java:67)
    Downloading file: https://78.media.tumblr.com/493c5e20a5d4ba7395ea830e2867cf06/tumblr_o6qxe6fGOa1vn8reqo1_r1_1280.png Retry #1
java.io.IOException: No such file or directory
	at java.io.UnixFileSystem.createFileExclusively(Native Method)
	at java.io.File.createNewFile(File.java:1012)
	at com.rarchives.ripme.ripper.AbstractRipper.writeDownloadedURL(AbstractRipper.java:77)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:256)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:265)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:269)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:283)
	at com.rarchives.ripme.ripper.AlbumRipper.addURLToDownload(AlbumRipper.java:112)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.handleJSON(TumblrRipper.java:237)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.rip(TumblrRipper.java:191)
	at com.rarchives.ripme.App.rip(App.java:92)
	at com.rarchives.ripme.App.ripURL(App.java:250)
	at com.rarchives.ripme.App.handleArguments(App.java:237)
	at com.rarchives.ripme.App.main(App.java:67)
    Downloading file: https://78.media.tumblr.com/9fd400a000d12745d76c4bf704a8762e/tumblr_o6qxe6fGOa1vn8reqo2_r2_1280.jpg Retry #1
java.io.IOException: No such file or directory
	at java.io.UnixFileSystem.createFileExclusively(Native Method)
	at java.io.File.createNewFile(File.java:1012)
	at com.rarchives.ripme.ripper.AbstractRipper.writeDownloadedURL(AbstractRipper.java:77)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:256)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:265)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:269)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:283)
	at com.rarchives.ripme.ripper.AlbumRipper.addURLToDownload(AlbumRipper.java:112)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.handleJSON(TumblrRipper.java:237)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.rip(TumblrRipper.java:191)
	at com.rarchives.ripme.App.rip(App.java:92)
	at com.rarchives.ripme.App.ripURL(App.java:250)
	at com.rarchives.ripme.App.handleArguments(App.java:237)
	at com.rarchives.ripme.App.main(App.java:67)
Using api key: qpdkY6nMknksfvYAhf2xIHp0iNRLkMlcWShxqzXyFJRxIsZ1Zz
Retrieving http://api.tumblr.com/v2/blog/momentummod.tumblr.com/posts/photo?api_key=qpdkY6nMknksfvYAhf2xIHp0iNRLkMlcWShxqzXyFJRxIsZ1Zz&offset=20
    Downloading file: https://78.media.tumblr.com/25030f041b0e6c6d4276fd2614df3e5c/tumblr_o6qxe6fGOa1vn8reqo3_r1_1280.png Retry #1
[+] Saved https://78.media.tumblr.com/9fd400a000d12745d76c4bf704a8762e/tumblr_o6qxe6fGOa1vn8reqo2_r2_1280.jpg as ./rips/tumblr_momentummod.tumblr.com/tumblr_o6qxe6fGOa1vn8reqo2_r2_1280.jpg
[+] Saved https://78.media.tumblr.com/25030f041b0e6c6d4276fd2614df3e5c/tumblr_o6qxe6fGOa1vn8reqo3_r1_1280.png as ./rips/tumblr_momentummod.tumblr.com/tumblr_o6qxe6fGOa1vn8reqo3_r1_1280.png
[+] Saved https://78.media.tumblr.com/493c5e20a5d4ba7395ea830e2867cf06/tumblr_o6qxe6fGOa1vn8reqo1_r1_1280.png as ./rips/tumblr_momentummod.tumblr.com/tumblr_o6qxe6fGOa1vn8reqo1_r1_1280.png
   Zero posts returned.
Using api key: qpdkY6nMknksfvYAhf2xIHp0iNRLkMlcWShxqzXyFJRxIsZ1Zz
Retrieving http://api.tumblr.com/v2/blog/momentummod.tumblr.com/posts/video?api_key=qpdkY6nMknksfvYAhf2xIHp0iNRLkMlcWShxqzXyFJRxIsZ1Zz&offset=0
   Zero posts returned.

~/Pictures
❯ 
```
</details>

**Comment by cyian-1756** [comment]
What dir are you saving rips to? Does it exist?

**Comment by bored** [comment]
@cyian-1756 I was under the impression RipMe created any directories that didn't already exist, but I tested it and it doesn't seem to be the cause.

Even if I run the `java -jar ripme.jar -u http://momentummod.tumblr.com` command again when all the directories and folders have been created during the first rip, the same set of exceptions occur.

<details>
<summary>Full Output</summary>

```
~/Backups
❯ mkdir -p rips/tumblr_momentummod.tumblr.com

~/Backups
❯ java -jar ripme.jar -u http://momentummod.tumblr.com
Loaded file:/home/adam/Backups/ripme.jar!/rip.properties
Loaded log4j.properties
Initialized ripme v1.7.47
Using api key: FQrwZMCxVnzonv90rgNUJcAk4FpnoS0mYuSuGYqIpM2cFgp9L4
Retrieving http://api.tumblr.com/v2/blog/momentummod.tumblr.com/posts/photo?api_key=FQrwZMCxVnzonv90rgNUJcAk4FpnoS0mYuSuGYqIpM2cFgp9L4&offset=0
java.io.IOException: No such file or directory
	at java.io.UnixFileSystem.createFileExclusively(Native Method)
	at java.io.File.createNewFile(File.java:1012)
	at com.rarchives.ripme.ripper.AbstractRipper.writeDownloadedURL(AbstractRipper.java:77)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:256)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:265)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:269)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:283)
	at com.rarchives.ripme.ripper.AlbumRipper.addURLToDownload(AlbumRipper.java:112)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.handleJSON(TumblrRipper.java:237)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.rip(TumblrRipper.java:191)
	at com.rarchives.ripme.App.rip(App.java:92)
	at com.rarchives.ripme.App.ripURL(App.java:250)
	at com.rarchives.ripme.App.handleArguments(App.java:237)
	at com.rarchives.ripme.App.main(App.java:67)
    Downloading file: https://78.media.tumblr.com/493c5e20a5d4ba7395ea830e2867cf06/tumblr_o6qxe6fGOa1vn8reqo1_r1_1280.png Retry #1
java.io.IOException: No such file or directory
	at java.io.UnixFileSystem.createFileExclusively(Native Method)
	at java.io.File.createNewFile(File.java:1012)
	at com.rarchives.ripme.ripper.AbstractRipper.writeDownloadedURL(AbstractRipper.java:77)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:256)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:265)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:269)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:283)
	at com.rarchives.ripme.ripper.AlbumRipper.addURLToDownload(AlbumRipper.java:112)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.handleJSON(TumblrRipper.java:237)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.rip(TumblrRipper.java:191)
	at com.rarchives.ripme.App.rip(App.java:92)
	at com.rarchives.ripme.App.ripURL(App.java:250)
	at com.rarchives.ripme.App.handleArguments(App.java:237)
	at com.rarchives.ripme.App.main(App.java:67)
    Downloading file: https://78.media.tumblr.com/9fd400a000d12745d76c4bf704a8762e/tumblr_o6qxe6fGOa1vn8reqo2_r2_1280.jpg Retry #1
java.io.IOException: No such file or directory
	at java.io.UnixFileSystem.createFileExclusively(Native Method)
	at java.io.File.createNewFile(File.java:1012)
	at com.rarchives.ripme.ripper.AbstractRipper.writeDownloadedURL(AbstractRipper.java:77)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:256)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:265)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:269)
	at com.rarchives.ripme.ripper.AbstractRipper.addURLToDownload(AbstractRipper.java:283)
	at com.rarchives.ripme.ripper.AlbumRipper.addURLToDownload(AlbumRipper.java:112)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.handleJSON(TumblrRipper.java:237)
	at com.rarchives.ripme.ripper.rippers.TumblrRipper.rip(TumblrRipper.java:191)
	at com.rarchives.ripme.App.rip(App.java:92)
	at com.rarchives.ripme.App.ripURL(App.java:250)
	at com.rarchives.ripme.App.handleArguments(App.java:237)
	at com.rarchives.ripme.App.main(App.java:67)
Using api key: FQrwZMCxVnzonv90rgNUJcAk4FpnoS0mYuSuGYqIpM2cFgp9L4
Retrieving http://api.tumblr.com/v2/blog/momentummod.tumblr.com/posts/photo?api_key=FQrwZMCxVnzonv90rgNUJcAk4FpnoS0mYuSuGYqIpM2cFgp9L4&offset=20
    Downloading file: https://78.media.tumblr.com/25030f041b0e6c6d4276fd2614df3e5c/tumblr_o6qxe6fGOa1vn8reqo3_r1_1280.png Retry #1
[+] Saved https://78.media.tumblr.com/9fd400a000d12745d76c4bf704a8762e/tumblr_o6qxe6fGOa1vn8reqo2_r2_1280.jpg as ./rips/tumblr_momentummod.tumblr.com/tumblr_o6qxe6fGOa1vn8reqo2_r2_1280.jpg
[+] Saved https://78.media.tumblr.com/25030f041b0e6c6d4276fd2614df3e5c/tumblr_o6qxe6fGOa1vn8reqo3_r1_1280.png as ./rips/tumblr_momentummod.tumblr.com/tumblr_o6qxe6fGOa1vn8reqo3_r1_1280.png
[+] Saved https://78.media.tumblr.com/493c5e20a5d4ba7395ea830e2867cf06/tumblr_o6qxe6fGOa1vn8reqo1_r1_1280.png as ./rips/tumblr_momentummod.tumblr.com/tumblr_o6qxe6fGOa1vn8reqo1_r1_1280.png
   Zero posts returned.
Using api key: FQrwZMCxVnzonv90rgNUJcAk4FpnoS0mYuSuGYqIpM2cFgp9L4
Retrieving http://api.tumblr.com/v2/blog/momentummod.tumblr.com/posts/video?api_key=FQrwZMCxVnzonv90rgNUJcAk4FpnoS0mYuSuGYqIpM2cFgp9L4&offset=0
   Zero posts returned.

~/Backups
❯ 
```
</details>

**Comment by cyian-1756** [comment]
@bored 

It looks like the error is getting thrown because ripme can't write the url history file. 

Run `mkdir ~/.config/ripme && touch  ~/.config/ripme/url_history.txt` and try downloading again

**Comment by cyian-1756** [comment]
@bored 

After some testing I can reproduce this if I delete the ~/.config/ripme folder or changes the permissions so that ripme can't write to it.

**Comment by bored** [comment]
@cyian-1756 glad you figured it out, I've actually been wondering why `rip.properties` wasn't generating for me. Is this a Java-related thing? I've never had problems with programs auto-generating things inside the `~/.config` directory before.

**Comment by cyian-1756** [comment]
>  Is this a Java-related thing?

I honestly don't know. I've never seen this issue popup before. Does debian have some weird permissions on ~/.config or something?

**Comment by bored** [comment]
@cyian-1756 I don't think so, I tested on Ubuntu 18.04 and had the exact same issue. Is it possible this only happens when you exclusively use the CLI? Just guessing at this point.

I have a ton of other programs that have generated configs in there, the permissions are `755` as you'd expect.

**Comment by cyian-1756** [comment]
> Is it possible this only happens when you exclusively use the CLI?

That could be it. I've only done testing from the GUI

**Comment by cyian-1756** [comment]
@bored Testing from a tty it seems that this issue does only appear when there ripme is being from the command line

**Comment by bored** [comment]
@cyian-1756 glad you figured it out, I did manage to test it out in visual mode and it seems the config is only created during `shutdownCleanup()`.

Also, I'm not really too clued up on Java but I had a quick look at the only other place `saveConfig()` can be called from (inside `ripURL`) and it doesn't seem right to me, but maybe I'm misunderstanding. I tested it out and the config is only created/saved when you pass the `-n, --no-prop-file` but surely that is backwards?
